src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -0,0 +1,264 @@
+
+        protected override string GetListenerName() => DiagnosticListenerName;
+
+        protected override void OnNext(string eventName, object arg)
+        {
+            switch (eventName)
+            {
+                case "Microsoft.AspNetCore.Hosting.HttpRequestIn.Start":
+                    OnHostingHttpRequestInStart(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Hosting.UnhandledException":
+                    OnHostingUnhandledException(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Hosting.HttpRequestIn.Stop":
+                    OnHostingHttpRequestInStop(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Mvc.BeforeAction":
+                    OnMvcBeforeAction(arg);
+                    break;
+
+                case "Microsoft.AspNetCore.Mvc.AfterOnActionExecuted":
+                    OnMvcAfterOnActionExecuted(arg);
+                    break;
+            }
+        }
+
+        private static string GetUrl(HttpRequest request)
+        {
+            if (request.Host.HasValue)
+            {
+                return $"{request.Scheme}://{request.Host.Value}{request.PathBase.Value}{request.Path.Value}";
+            }
+
+            // HTTP 1.0 requests are not required to provide a Host to be valid
+            // Since this is just for display, we can provide a string that is
+            // not an actual Uri with only the fields that are specified.
+            // request.GetDisplayUrl(), used above, will throw an exception
+            // if request.Host is null.
+            return $"{request.Scheme}://{NoHostSpecified}{request.PathBase.Value}{request.Path.Value}";
+        }
+
+        private static SpanContext ExtractPropagatedContext(HttpRequest request)
+        {
+            try
+            {
+                // extract propagation details from http headers
+                var requestHeaders = request.Headers;
+
+                if (requestHeaders != null)
+                {
+                    var headersCollection = new DictionaryHeadersCollection();
+
+                    foreach (var header in requestHeaders)
+                    {
+                        string key = header.Key;
+                        string[] values = header.Value.ToArray();
+
+                        if (key != null && values.Length > 0)
+                        {
+                            headersCollection.Add(key, values);
+                        }
+                    }
+
+                    return SpanContextPropagator.Instance.Extract(headersCollection);
+                }
+            }
+            catch (Exception ex)
+            {
+                if (Log.IsEnabled(LogEventLevel.Error))
+                {
+                    Log.Error(ex, "Error extracting propagated HTTP headers.");
+                }
+            }
+
+            return null;
+        }
+
+        private void OnMvcAfterOnActionExecuted(object arg)
```

Comment:
nit: I would find it helpful if the methods were ordered in the expected execution order, could you do that?

Reply:
I agree. Would be great.

Reply:
Changed in a229a5c839aa838cf710eb19ca3fb5c99519b289.